[intesis] - added IntesisBox support

hmerk opened the pull request:
Signed-off-by: Hans-Jörg Merk <github@hmerk.de>
Also Signed-off-by: Rocky Amatulli <rocky@amatulli.net>
Also Signed-off-by: Cody Cutrer <cody@cutrer.us>

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/IntesisConfiguration.java:
```diff
@@ -23,4 +23,5 @@
 public class IntesisConfiguration {
```
Please split this in two classes if the Things have different configurations.

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxIdentity.java:
```diff
@@ -0,0 +1,37 @@
+@NonNullByDefault
+public class IntesisBoxIdentity {
+
+    public HashMap<String, String> value = new HashMap<String, String>();
```
```suggestion
    public HashMap<String, String> value = new HashMap<>();
```

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:
```diff
@@ -0,0 +1,232 @@
+            logger.debug("IntesisBox listener added");
+        }
+
+        tcpSocket = new IntesisSocket();
```
You could do this, to avoid the null warning.
```suggestion
        IntesisSocket localSocket = tcpSocket = new IntesisSocket();
```

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:
```diff
@@ -0,0 +1,232 @@
+    @SuppressWarnings("null")
+    public void closeConnection() {
+        try {
+            if (tcpSocket != null) {
```
The field can be null at this point, because it isn't inittialized in the constructor. So, you need to check if it is not null before accessing it: if (job != null).But this would result in a compiler warning, as it is a field and could theoretically be modified by another thread exactly between the null check and the access in the next line.To mitigate this, you can store the field to a local variable before checking it:
```

Future<?> localPollingJob = pollingJob;
if (localPollingJob != null) {
    localPollingJob.cancel(true);
}
```

hmerk replied:
Suppress Warning "null" is not needed any more.

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:
```diff
@@ -0,0 +1,232 @@
+    public void closeConnection() {
+        try {
+            if (tcpSocket != null) {
+                logger.debug("closeConnection(): Closing Socket!");
```
Can this message be replaced by using the debugger or by increasing the framework's log level? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

hmerk replied:
Remove

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:
```diff
@@ -0,0 +1,232 @@
+                    try {
+                        logger.trace("Calling readMessage()");
+                        readMessage(message);
+                    } catch (Exception e) {
```
What type of exception do you expect here? Can you specify the concrete type? Same for below.

hmerk replied:
removed try catch here

hmerk replied:
removed try catch here

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:
```diff
@@ -0,0 +1,232 @@
+        } catch (IOException ioException) {
+            logger.debug("write(): {}", ioException.getMessage());
+            setConnected(false);
+        } catch (Exception exception) {
```
See above

hmerk replied:
removed second catch

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:
```diff
@@ -0,0 +1,232 @@
+        } catch (IOException ioException) {
+            logger.debug("read(): IO Exception: {}", ioException.getMessage());
+            setConnected(false);
+        } catch (Exception exception) {
```
See above

hmerk replied:
removed second catch

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:
```diff
@@ -0,0 +1,232 @@
+                tcpOutput.flush();
+            }
+        } catch (IOException ioException) {
+            logger.debug("write(): {}", ioException.getMessage());
```
Can you update the Thing status with the exception's message? Same for the other occurrences.

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:
```diff
@@ -0,0 +1,232 @@
+ * @author Hans-Jörg Merk - Moved Socket to it's own class
+ */
+@NonNullByDefault
+public class IntesisBoxSocketApi {
```
Please significantly reduce the logging in this file. You can use the debugger to debug the code.

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,371 @@
+
+    private IntesisConfiguration config = new IntesisConfiguration();
+
+    private double minTemp = 0.0d, maxTemp = 0.0d;
```
Decimal numbers with point are `double` by default.
```suggestion
    private double minTemp = 0.0, maxTemp = 0.0;
```

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,371 @@
+                }
+            }
+            api.sendAlive();
+            ;
```
```suggestion
```

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,371 @@
+                return;
+            }
+            if (command instanceof RefreshType) {
+                logger.warn("Refresh channel {}", channelUID.getId());
```
This could be trace.

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,371 @@
+            updateStatus(ThingStatus.UNKNOWN);
+            scheduler.submit(() -> {
+                try {
+                    intesisBoxSocketApi.openConnection();
+                    intesisBoxSocketApi.sendId();
+                    intesisBoxSocketApi.sendLimitsQuery();
+                    intesisBoxSocketApi.sendAlive();
```
Please handle the null compiler warning.

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,371 @@
+        switch (channelUID.getId()) {
+            case CHANNEL_TYPE_POWER:
+                if (command instanceof OnOffType) {
+                    value = ((OnOffType) command == OnOffType.ON) ? "ON" : "OFF";
```
```suggestion
                    value = command == OnOffType.ON ? "ON" : "OFF";
```

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,371 @@
+                    celsiusTemperature = celsiusTemperature.toUnit(SIUnits.CELSIUS);
+                    if (celsiusTemperature != null) {
+                        double doubleValue = celsiusTemperature.doubleValue();
+                        if (doubleValue < minTemp) {
+                            doubleValue = minTemp;
+                        } else if (doubleValue > maxTemp) {
+                            doubleValue = maxTemp;
+                        }
```
You could use the Math function.
```suggestion
                        doubleValue = Math.max(minTemp, Math.min(maxTemp, doubleValue));
```

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,371 @@
+                        } else if (doubleValue > maxTemp) {
+                            doubleValue = maxTemp;
+                        }
+                        value = String.valueOf((int) Math.round(doubleValue * 10));
```
```suggestion
                        value = String.format("%.0f", doubleValue * 10);
```

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,371 @@
+                logger.trace("Sending command {} to function {}", value, function);
+                api.sendCommand(function, value);
+            } else {
+                logger.trace("Sending command failed, could not get API");
```
This could be warn

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,371 @@
+
+            case "SETPTEMP":
+                if (value.equals("32768")) {
+                    value = "0";
```
It's bad practice to modify method arguments. That's why the compiler complains. Better create a new local variable for it.

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,371 @@
+                return;
+            }
+        }
+        IntesisBoxMessage message = IntesisBoxMessage.parse(data);
+        switch (message.getCommand()) {
```
`parse()` returns Nullable. You could declare its return value as NonNull or do a null check here.

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,371 @@
+                List<StateOption> options = new ArrayList<>();
+                for (String mode : limits.get(channelId)) {
+                    options.add(new StateOption(mode,
+                            (mode.substring(0, 1).toUpperCase() + mode.substring(1).toLowerCase())));
```
```suggestion
                            mode.substring(0, 1).toUpperCase() + mode.substring(1).toLowerCase()));
```

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/README.md:
```diff
@@ -1,62 +1,65 @@
+| Channel ID         | Item Type          | Description                                 | Possible Values             |
+|--------------------|--------------------|---------------------------------------------|-----------------------------|
+| power              | Switch             | Turns power on/off for your climate system. | ON, OFF                     |
+| mode               | String             | The heating/cooling mode.                   | AUTO,HEAT,DRY,FAN,COOL      |
```
Is the "if applicable" missing for the mode Channel and some others?

hmerk replied:
Done

hmerk commented:
Thanks @fwolter 
I have addressed all your comments so far.

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxIdentity.java:
```diff
@@ -0,0 +1,37 @@
+ * @author Rocky Amatulli - Initial contribution
+ */
+@NonNullByDefault
+public class IntesisBoxIdentity {
```
I don't see why you need this class. Just replace it with a utility function.

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxMessage.java:
```diff
@@ -0,0 +1,79 @@
+    public static final String DISCOVER = "DISCOVER";
+
+    private static final Pattern REGEX = Pattern.compile(
+            "^(ID|INFO|SET|CHN|GET|LOGIN|LOGOUT|CFG|LIMITS)(?:,(\\d+))?:(APPVERSION|RUNVERSION|CFGVERSION|HASH|ONOFF|MODE|SETPTEMP|FANSP|VANEUD|VANELR|AMBTEMP|ERRSTATUS|ERRCODE),([A-Z0-9.,\\[\\]]+)$");
```
It would be better to parse out the command and function names in their entirety then check that the parsed values are acceptable. That way when you find a mismatch you can log the unexpected value vs not knowing what went wrong.
```suggestion
            "^([^,]+)(?:,(\\d+))?:([^,]+),([A-Z0-9.,\\[\\]]+)$");
```

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:
```diff
@@ -0,0 +1,199 @@
+        try {
+            if (tcpOutput != null) {
+                tcpOutput.write(data);
+            }
+            if (tcpOutput != null) {
```
You shouldn't need to check if `tcpOutput != null` twice.
```suggestion
```

hmerk replied:
If not checked again, Eclipse gives me a potential null pointer warning.

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:
```diff
@@ -0,0 +1,199 @@
+        IntesisBoxChangeListener listener = this.changeListener;
+        IntesisSocket localSocket = tcpSocket = new IntesisSocket();
+        tcpOutput = new OutputStreamWriter(localSocket.socket.getOutputStream(), "US-ASCII");
+        tcpInput = new BufferedReader(new InputStreamReader(localSocket.socket.getInputStream()));
```
You should specify a character encoding for the InputStreamReader.

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:
```diff
@@ -0,0 +1,199 @@
+        tcpOutput = new OutputStreamWriter(localSocket.socket.getOutputStream(), "US-ASCII");
+        tcpInput = new BufferedReader(new InputStreamReader(localSocket.socket.getInputStream()));
+
+        Thread tcpListener = new Thread(new TCPListener());
+        tcpListener.start();
```
Please give the thread a name that allows debuggers to easily identify this binding as the creator of it.
Also please make the thread daemon.

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:
```diff
@@ -0,0 +1,199 @@
+        this.port = port;
+    }
+
+    private class IntesisSocket {
```
Why is this class needed? It doesn't add any functionality so just keep track of the Socket instead.

hmerk replied:
That's how I saw it in the energenie Binding, which socket class was written by @Hilbrand .  I have no clue of socket handling, so it was simply copy and paste.

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,383 @@
+                break;
+        }
+        if (!value.isEmpty()) {
+            String function = IntesisBoxFunctionEnum.valueOf(channelUID.getId()).getFunction();
```
If you are only using this enum for one-directional lookups, it would be better to define the `function` as part of the switch statement above. That way you can remove the use of the enum entirely.

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,383 @@
+                updateState(CHANNEL_TYPE_VANESLR, new StringType(value));
+                break;
+            case "ERRCODE":
+                properties.put("errorCode", value);
+                updateProperties(properties);
+                break;
+            case "ERRSTATUS":
+                properties.put("errorStatus", value);
```
Why are errorCode and errorStatus properties instead of channels?

hmerk replied:
Changed to be channels

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,383 @@
+            switch (message.getCommand()) {
+                case LIMITS:
+                    logger.debug("handleMessage(): Limits received - {}", data);
+                    synchronized (this) {
```
Why are you synchronizing?

hmerk replied:
Removed

cpmeister commented:
Also please address the build warnings:
```
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:[91,17] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:[95,17] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:[99,17] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:[134,17] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:[137,17] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:[152,27] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:[96,13] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:[102,25] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:[105,25] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:[108,25] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] /home/travis/build/openhab/openhab-addons/bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:[111,25] Potential null pointer access: this expression has a '@Nullable' type
```

hmerk commented:
Thanks @cpmeister for your review, I have addressed most of it, some comments left.
According to the compiler warnings, i have already added null checks, but the warnings still exist. Don't have a clue how to fix this.

hmerk commented:
Failing build is not related to intesis binding but Feed tests...

cpmeister commented:
> According to the compiler warnings, i have already added null checks, but the warnings still exist. Don't have a clue how to fix this.

The way to fix them is to assign the `@Nullable` fields to a local variable and then do all your processing on that local variable instead. The null checker is operating under the assumption that a `@Nullable` field could be changed to null at any time by another thread.

hmerk commented:
Thanks @cpmeister I have fixed the compiler warnings and the double null check.

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -203,32 +217,17 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+        properties.put("ipAddress", value[2]);
+        properties.put("protocol", value[3]);
+        properties.put(PROPERTY_FIRMWARE_VERSION, value[4]);
+        properties.put("rssi", value[5]);
```
Doesn't rssi change over time? If it changes over time it should be a channel, not a property

hmerk replied:
Will change that later, as I have to think how to block other properties from regular updates. Should be no big issue.

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,370 @@
+                } catch (IOException e) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+                }
+                pollingTask = scheduler.scheduleWithFixedDelay(this::polling, 3, 45, TimeUnit.SECONDS);
+                updateStatus(ThingStatus.ONLINE);
+            });
```
Unless there are some kind of strict timing requirements, it would be better to start periodic tasks synchronously in `initialize()`.
```suggestion
                updateStatus(ThingStatus.ONLINE);
            });
            pollingTask = scheduler.scheduleWithFixedDelay(this::polling, 3, 45, TimeUnit.SECONDS);
```

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,370 @@
+
+    @Override
+    public void dispose() {
+        final ScheduledFuture<?> pollingTask = this.pollingTask;
+
+        IntesisBoxSocketApi api = this.intesisBoxSocketApi;
+
+        if (pollingTask != null) {
+            pollingTask.cancel(true);
+            this.pollingTask = null;
+        }
+        if (api != null) {
+            api.closeConnection();
+        }
```
Is there somewhere here you are calling `intesisBoxSocketApi.removeIntesisBoxChangeListener(this);`? If you aren't then you should.

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,370 @@
+
+    private IntesisBoxConfiguration config = new IntesisBoxConfiguration();
+
+    private double minTemp = 0.0, maxTemp = 0.0;
```
Should these values also be reset as part of `initialize()`?

hmerk replied:
No, as they get updated in the polling task.

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,370 @@
+                        intesisLocalApi.sendAlive();
+                    }
+                } catch (IOException e) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+                }
```
You should return here otherwise your thing status will be set to ONLINE immediately after.
```suggestion
                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
                    return;
                }
```

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,370 @@
+                addChannel(CHANNEL_TYPE_ERRORCODE, "String");
+                addChannel(CHANNEL_TYPE_ERRORSTATUS, "String");
+
+                intesisBoxSocketApi = new IntesisBoxSocketApi(config.ipAddress, config.port);
+                intesisBoxSocketApi.addIntesisBoxChangeListener(this);
+                try {
+                    IntesisBoxSocketApi intesisLocalApi = intesisBoxSocketApi;
+                    if (intesisLocalApi != null) {
+                        intesisLocalApi.openConnection();
+                        intesisLocalApi.sendId();
+                        intesisLocalApi.sendLimitsQuery();
+                        intesisLocalApi.sendAlive();
+                    }
```
If you assign the local variable at the same time you don't need to do null checks later on.
```suggestion
                IntesisBoxSocketApi intesisLocalApi = intesisBoxSocketApi = new IntesisBoxSocketApi(config.ipAddress, config.port);
                intesisLocalApi.addIntesisBoxChangeListener(this);
                try {
                    intesisLocalApi.openConnection();
                    intesisLocalApi.sendId();
                    intesisLocalApi.sendLimitsQuery();
                    intesisLocalApi.sendAlive();
                    
```

hmerk replied:
Done

hmerk commented:
Thanks @cpmeister I have addressed your comments, the rssi channel will be added later today.

hmerk commented:
@cpmeister RSSI is now changed from property to be a system-channel.

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/IntesisHandlerFactory.java:
```diff
@@ -73,6 +77,11 @@ public boolean supportsThingType(ThingTypeUID thingTypeUID) {
         }
 
+        if (THING_TYPE_INTESISBOX.equals(thingTypeUID)) {
+            logger.debug("Creating a IntesisBoxHandler for thing '{}'", thing.getUID());
```
This is logged by the framework an can therefore be removed.

hmerk replied:
removed

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:
```diff
@@ -0,0 +1,210 @@
+
+        IntesisBoxChangeListener listener = this.changeListener;
+        IntesisSocket localSocket = tcpSocket = new IntesisSocket();
+        tcpOutput = new OutputStreamWriter(localSocket.socket.getOutputStream(), "US-ASCII");
+        tcpInput = new BufferedReader(new InputStreamReader(localSocket.socket.getInputStream(), "US-ASCII"));
```
```suggestion
        tcpOutput = new OutputStreamWriter(localSocket.socket.getOutputStream(), StandardCharsets.US_ASCII);
        tcpInput = new BufferedReader(
                new InputStreamReader(localSocket.socket.getInputStream(), StandardCharsets.US_ASCII));
```

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisBoxSocketApi.java:
```diff
@@ -0,0 +1,210 @@
+        tcpInput = new BufferedReader(new InputStreamReader(localSocket.socket.getInputStream(), "US-ASCII"));
+
+        Thread tcpListener = new Thread(new TCPListener());
+        tcpListener.setName("IntesisBoxTCPListener");
```
Please use the naming scheme mentioned here #8216 .

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,395 @@
+    private @Nullable IntesisBoxSocketApi intesisBoxSocketApi;
+
+    private final Map<String, String> properties = new HashMap<>();
+    private final Map<String, List<String>> limits = new HashMap<String, List<String>>();
```
Syntactical sugar
```suggestion
    private final Map<String, List<String>> limits = new HashMap<>();
```

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisBoxHandler.java:
```diff
@@ -0,0 +1,395 @@
+
+            updateStatus(ThingStatus.UNKNOWN);
+            scheduler.submit(() -> {
+                addChannel(CHANNEL_TYPE_AMBIENTTEMP, "Number:Temperature");
+                addChannel(CHANNEL_TYPE_ERRORCODE, "String");
+                addChannel(CHANNEL_TYPE_ERRORSTATUS, "String");
```
If they are always added, can't they be specified in the XML file?

hmerk replied:
Done

hmerk commented:
Thanks @fwolter all comments addressed.

fwolter reviewed (approved):
LGTM

cpmeister reviewed (approved):
LGTM

hmerk commented:
Thanks @fwolter  and @cpmeister  for your great support !!!
